rust/lance-encoding/src/buffer.rs

```diff
@@ -114,6 +173,110 @@ impl LanceBuffer {
+    /// > (`T` having a less strict alignment is not sufficient, the alignment really
+    /// > needs to be equal to satisfy the [`dealloc`] requirement that memory must be
+    /// > allocated and deallocated with the same layout.)
+    ///
+    /// However, we can safely reinterpret Vec<T> into &[u8] which is what happens here.
+    pub fn reinterpret_vec<T: ArrowNativeType>(vec: Vec<T>) -> Self {
+        Self::Borrowed(Buffer::from_vec(vec))
+    }
+
+    /// Reinterprets a LanceBuffer into a Vec<T>
+    ///
+    /// Unfortunately, there is no way to do this safely in Rust without a copy, even if
+    /// the source is Vec<u8>.
+    pub fn borrow_to_typed_slice<T: ArrowNativeType>(&mut self) -> impl AsRef<[T]> {
+        ScalarBuffer::<T>::from(self.borrow_and_clone().into_buffer())
+    }
+
+    /// Concatenates multiple buffers into a single buffer, consuming the input buffers
+    ///
+    /// If there is only one buffer, it will be returned as is
+    pub fn concat_into_one(buffers: Vec<Self>) -> Self {
+        if buffers.len() == 1 {
+            return buffers.into_iter().next().unwrap();
+        }
+
+        let mut total_len = 0;
+        for buffer in &buffers {
+            total_len += buffer.len();
+        }
+
+        let mut data = Vec::with_capacity(total_len);
+        for buffer in buffers {
+            data.extend_from_slice(buffer.as_ref());
+        }
+
+        Self::Owned(data)
+    }
+
+    /// Zips multiple buffers into a single buffer, consuming the input buffers
+    ///
+    /// Unlike concat_into_one this "zips" the buffers, interleaving the values
+    pub fn zip_into_one(buffers: Vec<(Self, u64)>, num_values: u64) -> Result<Self> {
+        let bytes_per_value = buffers.iter().map(|(_, bits_per_value)| {
+            if bits_per_value % 8 == 0 {
+                Ok(bits_per_value / 8)
+            } else {
+                Err(Error::InvalidInput { source: format!("LanceBuffer::zip_into_one only supports full-byte buffers currently and received a buffer with {} bits per value", bits_per_value).into(), location: location!() })
+            }
+        }).collect::<Result<Vec<_>>>()?;
+        let total_bytes_per_value = bytes_per_value.iter().sum::<u64>();
+        let total_bytes = (total_bytes_per_value * num_values) as usize;
+
+        let mut zipped = vec![0_u8; total_bytes];
+        let mut buffer_ptrs = buffers
+            .iter()
+            .zip(bytes_per_value)
+            .map(|((buffer, _), bytes_per_value)| (buffer.as_ptr(), bytes_per_value as usize))
+            .collect::<Vec<_>>();
+
+        let mut zipped_ptr = zipped.as_mut_ptr();
+        unsafe {
+            let end = zipped_ptr.add(total_bytes);
+            while zipped_ptr < end {
+                for (buf, bytes_per_value) in buffer_ptrs.iter_mut() {
+                    std::ptr::copy_nonoverlapping(*buf, zipped_ptr, *bytes_per_value);
+                    zipped_ptr = zipped_ptr.add(*bytes_per_value);
```

Comment:
[write_unaligned](https://doc.rust-lang.org/1.80.1/src/core/ptr/mod.rs.html#1620) saves some checking than [copy_nonoverlappiing](https://doc.rust-lang.org/1.80.1/src/core/intrinsics.rs.html#2934) and therefore faster when doing unaligned store

Reply:
It looks like `write_unaligned` just calls `copy_nonoverlapping`?  Also, I think to use `write_unaligned` I need to have `Vec<T>` and not `Vec<u8>` which is difficult to obtain.

However, if you can find any way to make this faster, feel free to investigate, but I don't think it is a priority yet (see below).  I added a benchmark.  Right now zipping buffers is very slow (1GBps).

Fast `zip` and `unzip` would be very useful.  Right now I think this will be the slowest part of the approach I'd like to eventually adopt.  However, `unzip` (which is on the read path) is more important to optimize than `zip` (which is on the write path).  Also, I think the important operation will not be "zip/unzip two byte-aligned buffers" but rather "zip/unzip two bit buffers" (zipping repetition levels, definition levels, and bitpacked values together).

So I think the important operations to investigate for performance are not ready yet.  Let's get further along and then we can drill down once we have things more complete.

Reply:
not exactly suit our use here, but some reference from [risingwave's implementation of zip_eq_fast](https://github.com/risingwavelabs/risingwave/pull/7795)

Reply:
std::ptr::copy_nonoverlapping:
```rust
pub const unsafe fn copy_nonoverlapping<T>(src: *const T, dst: *mut T, count: usize) {
    extern "rust-intrinsic" {
        #[rustc_const_unstable(feature = "const_intrinsic_copy", issue = "80697")]
        #[rustc_nounwind]
        pub fn copy_nonoverlapping<T>(src: *const T, dst: *mut T, count: usize);
    }

    ub_checks::assert_unsafe_precondition!(
        check_language_ub,
        "ptr::copy_nonoverlapping requires that both pointer arguments are aligned and non-null \
        and the specified memory ranges do not overlap",
        (
            src: *const () = src as *const (),
            dst: *mut () = dst as *mut (),
            size: usize = size_of::<T>(),
            align: usize = align_of::<T>(),
            count: usize = count,
        ) =>
        ub_checks::is_aligned_and_not_null(src, align)
            && ub_checks::is_aligned_and_not_null(dst, align)
            && ub_checks::is_nonoverlapping(src, dst, size, count)
    );

    // SAFETY: the safety contract for `copy_nonoverlapping` must be
    // upheld by the caller.
    unsafe { copy_nonoverlapping(src, dst, count) }
}
```
`std::ptr::copy_nonoverlapping` requires alignment but `write_unaligned` can directly use cpu's `unaligned store`, I remember I witnessed huge speed up when switching `std::ptr::copy_nonoverlapping` to `std::ptr::write_unaligned` in #2626.
just for future reference, let's punt it for now